base/server/src/main/java/com/netscape/cmscore/cert/CertUtils.java

```diff
@@ -569,6 +571,124 @@ public static String getCertType(X509CertImpl cert) throws CertificateParsingExc
         return (sb.length() > 0) ? sb.toString() : null;
     }
 
+    public static void addExtension(String name, Extension ext, X509CertInfo info)
+            throws EProfileException {
+        if (ext == null) {
+            throw new EProfileException("addExtension: extension '" + name + "' is null");
+        }
+        CertificateExtensions exts = null;
+
+        Extension alreadyPresentExtension = getExtension(name, info);
+
+        if (alreadyPresentExtension != null) {
+            String eName = ext.toString();
+            logger.error("Duplicate extension: " + eName);
+            throw new EProfileException(CMS.getUserMessage("CMS_PROFILE_DUPLICATE_EXTENSION", eName));
+        }
+ 
+        try {
+            exts = (CertificateExtensions)
+                    info.get(X509CertInfo.EXTENSIONS);
+        } catch (Exception e) {
+            logger.warn("EnrollDefault: " + e.getMessage(), e);
+        }
+        if (exts == null) {
+            throw new EProfileException("extensions not found");
+        }
+        try {
+            exts.set(name, ext);
+        } catch (IOException e) {
+            logger.warn("EnrollDefault: " + e.getMessage(), e);
+        }
+    }
+
+    public static void deleteExtension(String extID, X509CertInfo info) throws Exception {
+
+        CertificateExtensions exts = (CertificateExtensions) info.get(X509CertInfo.EXTENSIONS);
+
+        if (exts == null) {
+            return;
+        }
+
+        Collection<String> names = new ArrayList<>();
+        Enumeration<String> e = exts.getNames();
+
+        // get names of extensions to remove
+        while (e.hasMoreElements()) {
+            String name = e.nextElement();
+            Extension ext = (Extension) exts.get(name);
+
+            if (ext.getExtensionId().toString().equals(extID)) {
+                names.add(name);
+            }
+        }
+
+        // remove extensions in separate loop to avoid ConcurrentModificationException
+        for (String name : names) {
+            exts.delete(name);
+        }
+    }
+
+    public static void replaceExtension(String name, Extension ext, X509CertInfo info)
+            throws EProfileException {
+        try {
+            deleteExtension(name, info);
+        } catch (Exception e) {
+            throw new EProfileException(e);
+        }
+
+        addExtension(name, ext, info);
+    }
+
+    public static Extension getExtension(String name, X509CertInfo info) {
```

Comment:
Just a comment here on the returning of null if the input is not valid. I guess here the caller can't distinguish whether or not the the extension was not present or that the input was bad. Would it make sense to throw an exception in the bad input case?